arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -0,0 +1,826 @@
+  {
+    // Quick return if the stream isn't open.
+    if (!this->is_valid())
+    {
+      return static_cast<unsigned int>(-1);
+    }
+
+    return static_cast<unsigned int>(
+      (this->f_pts - this->f_start_time) / this->stream_time_base_to_frame()
+      - static_cast<int>(this->f_frame_number_offset));
+  }
+
+}; // end of internal class.
+
+// static open interlocking mutex
+std::mutex ffmpeg_video_input::priv::open_mutex;
+
+
+// ==================================================================
+ffmpeg_video_input
+::ffmpeg_video_input()
+  : d( new priv() )
+{
+  attach_logger( "ffmpeg_video_input" ); // get appropriate logger
+  d->logger = this->logger();
+  ffmpeg_init();
+}
+
+
+ffmpeg_video_input
+::~ffmpeg_video_input()
+{
+  this->close();
+}
+
+
+// ------------------------------------------------------------------
+// Get this algorithm's \link vital::config_block configuration block \endlink
+vital::config_block_sptr
+ffmpeg_video_input
+::get_configuration() const
+{
+  // get base config from base class
+  vital::config_block_sptr config = vital::algo::video_input::get_configuration();
+
+  return config;
+}
+
+
+// ------------------------------------------------------------------
+// Set this algorithm's properties via a config block
+void
+ffmpeg_video_input
+::set_configuration(vital::config_block_sptr in_config)
+{
+  // Starting with our generated vital::config_block to ensure that assumed values are present
+  // An alternative is to check for key presence before performing a get_value() call.
+
+  vital::config_block_sptr config = this->get_configuration();
+  config->merge_config(in_config);
+}
+
+
+// ------------------------------------------------------------------
+bool
+ffmpeg_video_input
+::check_configuration(vital::config_block_sptr config) const
+{
+  bool retcode(true); // assume success
+
+  return retcode;
+}
+
+
+// ------------------------------------------------------------------
+void
+ffmpeg_video_input
+::open( std::string video_name )
+{
+  // Close any currently opened file
+  this->close();
+
+  d->video_path = video_name;
+
+  {
+    std::lock_guard< std::mutex > lock(d->open_mutex);
+
+    if (!kwiversys::SystemTools::FileExists(d->video_path))
+    {
+      // Throw exception
+      throw kwiver::vital::file_not_found_exception(video_name, "File not found");
+    }
+
+    if (!d->open(video_name))
+    {
+      throw kwiver::vital::video_runtime_exception("Video stream open failed for unknown reasons");
+    }
+    d->end_of_video = false;
+  }
+}
+
+
+// ------------------------------------------------------------------
+void
+ffmpeg_video_input
+::close()
+{
+  if (d->f_packet.data) {
+    av_free_packet(&d->f_packet);  // free last packet
+  }
+
+  if (d->f_frame)
+  {
+    av_freep(&d->f_frame);
+  }
+  d->f_frame = nullptr;
+
+  if (d->f_video_encoding && d->f_video_encoding->opaque)
+  {
+    av_freep(&d->f_video_encoding->opaque);
+  }
+
+  //d->num_frames_ = -2;
```

Comment:
Remove this is it is no longer relevant nor is it like to be used in the future.